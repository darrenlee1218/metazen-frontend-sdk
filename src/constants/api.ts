export const API_CONSTANTS = Object.freeze({
  ASSET_POLLING_INTERVAL_S: 5,
  ASSET_POLLING_INTERVAL_MS: 5000,
  EVALUATOR_POLLING_INTERVAL_MS: 60000,
});
