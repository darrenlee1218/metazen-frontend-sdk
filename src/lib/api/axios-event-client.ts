import axios, { AxiosError, AxiosInstance, AxiosRequestConfig } from 'axios';
import EventEmitter from 'events';

export interface CustomApiClientOptions {
  enableErrorHandling?: boolean;
  transformErrorResponse?: (data: unknown) => string | null;
}

declare module 'axios' {
  interface AxiosRequestConfig extends CustomApiClientOptions {}
}

interface ApiEvent {
  'api:error': (message: string) => void;
  'api:unauthorized': () => void;
}

export declare interface AxiosEventClient {
  on: <U extends keyof ApiEvent>(event: U, listener: ApiEvent[U]) => this;
  off: <U extends keyof ApiEvent>(event: U, listener: ApiEvent[U]) => this;
  emit: <U extends keyof ApiEvent>(event: U, ...args: Parameters<ApiEvent[U]>) => boolean;
}

export const apiClientInstances: AxiosEventClient[] = [];

export class AxiosEventClient extends EventEmitter {
  private readonly _name: string;
  private readonly _client: AxiosInstance;
  private readonly _transformError: ((error: AxiosError) => string | null) | undefined;

  private readonly expected401Paths = new Set<string>(['/api/login/status', '/api/mfa/status']);

  constructor(name: string, config: AxiosRequestConfig, transformError?: (error: AxiosError) => string | null) {
    super();

    this._name = name;
    this._transformError = transformError;

    this._client = axios.create(config);
    this._client.interceptors.response.use((response) => response, this.onRejected.bind(this));

    apiClientInstances.push(this);
  }

  getClient(): AxiosInstance {
    return this._client;
  }

  private onRejected(error: AxiosError) {
    console.error({
      service: this._name,
      status: error.response?.status,
      data: error.response?.data,
    });

    try {
      const requestUrl = new URL(error.request.responseURL);
      if (error.response?.status === 401 && !this.expected401Paths.has(requestUrl.pathname)) {
        this.emit('api:unauthorized');
      }

      if (error.config.enableErrorHandling) {
        const message =
          error.config.transformErrorResponse?.(error.response?.data) ?? this._transformError?.(error) ?? error.message;

        this.emit('api:error', message);
      }
    } catch (e: unknown) {
      console.error(e);
    }

    throw error;
  }
}
