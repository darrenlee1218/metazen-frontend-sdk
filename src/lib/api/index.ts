import { kycApi } from './kyc';
import { userApi } from './user';

export const api = {
  kyc: kycApi,
  user: userApi,
};
