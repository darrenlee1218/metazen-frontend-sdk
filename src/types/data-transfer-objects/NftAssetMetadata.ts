import { NftCoreData } from './NftCoreData';

interface NftAssetMetadata {
  symbol: string;
  chainId: string;
  contractAddress: string;
  owner: string;
  tokenId: string;
  tokenStandard: string;
  createdTime: string;
  updatedTime: string;
  data: NftCoreData;
}

export default NftAssetMetadata;
