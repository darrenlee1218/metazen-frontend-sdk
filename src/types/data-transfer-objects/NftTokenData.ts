/**
 * @address address of current wallet
 * @assetId chainId_ticker_0x${contractAddress.substring(0, 8)}
 * @nftTokenId tokenId in the ERC721 / ERC1155 contract
 * @assetTicker deprecated
 */
interface NftTokenData {
  entityId: string;
  chainId: string;
  address: string;
  balance: string;
  assetId: string;
  nftTokenId: string;
  assetTicker: string;
}

export default NftTokenData;
