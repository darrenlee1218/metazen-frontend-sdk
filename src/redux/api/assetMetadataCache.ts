import { createApi } from '@reduxjs/toolkit/query/react';

import { jsonRPCRequestBaseQuery } from '@lib/RTKQueryJRPCBaseQuery';
import { assetMetadataCacheServiceClient } from '@services/APIServices/asset-metadata-cache-service';
import NftAssetMetadata from '@gryfyn-types/data-transfer-objects/NftAssetMetadata';

import { NftAssetMetadataResponse } from '@services/APIServices/asset-metadata-cache-service/NftAssetMetadataResponse';

type GetAssetMetadataBatchParams = GetAssetMetadataParams[];

export interface GetAssetMetadataParams {
  chainId: string | void;
  contractAddress: string | void;
  tokenId: string | void;
}

const transformAssetMetadataResponse = (response: NftAssetMetadataResponse): NftAssetMetadata => ({
  symbol: response.symbol ?? '',
  chainId: response.chainId ?? '',
  contractAddress: response.contractAddress ?? '',
  owner: response.owner ?? '',
  tokenId: response.tokenId ?? '',
  tokenStandard: response.tokenStandard ?? '',
  createdTime: response.createdTime ?? '',
  updatedTime: response.updatedTime ?? '',
  data: JSON.parse(response.data ?? '{}'),
});

export const assetMetadataCacheServiceApi = createApi({
  reducerPath: 'assetMetadataCacheService',
  baseQuery: jsonRPCRequestBaseQuery({ client: assetMetadataCacheServiceClient }),
  endpoints: (builder) => ({
    getAssetMetadata: builder.query<NftAssetMetadata | null, GetAssetMetadataParams>({
      query: (assetDetails) => ({ method: 'mtz_asset_metadata_cache_GetAssetMetadata', params: assetDetails }),
      transformResponse: (response: NftAssetMetadataResponse | null) => {
        if (response == null) return null;
        return transformAssetMetadataResponse(response);
      },
    }),

    refreshAssetMetadata: builder.query<null, GetAssetMetadataParams>({
      query: (assetDetails) => ({
        method: 'mtz_asset_metadata_cache_UpdateAssetMetadata',
        params: assetDetails,
      }),
    }),

    getAssetMetadataBatch: builder.query<NftAssetMetadata[] | null, GetAssetMetadataBatchParams>({
      query: (params) => ({ method: 'mtz_asset_metadata_cache_GetAssetMetadataBatch', params }),
      transformResponse: (response: NftAssetMetadataResponse[] | null) =>
        response?.map((res) => transformAssetMetadataResponse(res)) ?? null,
    }),
  }),
});

export const {
  useGetAssetMetadataQuery,
  useRefreshAssetMetadataQuery,
  useGetAssetMetadataBatchQuery,
  reducer: assetMetadataCacheServiceReducer,
  endpoints: assetMetadataCacheServiceEndpoints,
  middleware: assetMetadataCacheServiceMiddleware,
  reducerPath: assetMetadataCacheServiceReducerPath,
} = assetMetadataCacheServiceApi;
