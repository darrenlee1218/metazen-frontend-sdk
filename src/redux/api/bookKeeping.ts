import { createApi } from '@reduxjs/toolkit/query/react';

import { API_CONSTANTS } from '@constants/api';
import { jsonRPCRequestBaseQuery } from '@lib/RTKQueryJRPCBaseQuery';
import { bookkeepingClient } from '@services/APIServices/book-keeping';

import BalanceData from '@gryfyn-types/data-transfer-objects/BalanceData';
import NftTokenData from '@gryfyn-types/data-transfer-objects/NftTokenData';

export interface GetBalancesByAssetIdsParams {
  assetIds: string[];
  aggregateNfts?: boolean;
}

export const bookKeepingApi = createApi({
  reducerPath: 'bookKeeping',
  tagTypes: ['bookKeeping'],
  refetchOnMountOrArgChange: API_CONSTANTS.ASSET_POLLING_INTERVAL_S,
  baseQuery: jsonRPCRequestBaseQuery({ client: bookkeepingClient }),
  endpoints: (builder) => ({
    getBalances: builder.query<BalanceData[], void>({
      query: () => ({ method: 'mtz_bk_GetBalancesRequest' }),
    }),
    getBalancesByAssetIds: builder.query<NftTokenData[], GetBalancesByAssetIdsParams>({
      query: (assetDetails) => ({ method: 'mtz_bk_GetBalancesByAssetIdsRequest', params: assetDetails }),
    }),
  }),
});

export const {
  useGetBalancesQuery,
  useGetBalancesByAssetIdsQuery,
  reducer: bookKeepingReducer,
  middleware: bookKeepingMiddleware,
  reducerPath: bookKeepingReducerPath,
} = bookKeepingApi;

export const { getBalances } = bookKeepingApi.endpoints;
