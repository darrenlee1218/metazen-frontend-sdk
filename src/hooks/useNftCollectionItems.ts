import { API_CONSTANTS } from '@constants/api';
import NftAssetMetadata from '@gryfyn-types/data-transfer-objects/NftAssetMetadata';
import NftTokenData from '@gryfyn-types/data-transfer-objects/NftTokenData';
import { TransformedNftCollection } from '@gryfyn-types/data-transfer-objects/TransformedNftCollection';
import { useGetAssetMetadataBatchQuery } from '@redux/api/assetMetadataCache';
import { useGetBalancesByAssetIdsQuery } from '@redux/api/bookKeeping';
import BigNumber from 'bignumber.js';
import { useMemo } from 'react';

export interface NftItem extends NftTokenData {
  metadata?: NftAssetMetadata;
}

interface UseNftCollectionItemsResult {
  nftCollectionItems: NftItem[];
  isLoading: boolean;
  isLoadTokenError: boolean;
}

export const useNftCollectionItems = ({
  id: collectionKey,
  contractAddress,
}: Pick<TransformedNftCollection, 'id' | 'contractAddress'>): UseNftCollectionItemsResult => {
  const {
    data: nftTokens = [],
    isLoading: isLoadingTokenData,
    isError: isLoadTokenError,
  } = useGetBalancesByAssetIdsQuery(
    {
      assetIds: [collectionKey],
    },
    {
      pollingInterval: API_CONSTANTS.ASSET_POLLING_INTERVAL_MS,
    },
  );

  const filteredNftTokens = useMemo(
    () => nftTokens.filter((token) => new BigNumber(token.balance).isGreaterThan(0)),
    [nftTokens],
  );

  const { data: nftItemMetadata, isLoading: isLoadingMetadata } = useGetAssetMetadataBatchQuery(
    filteredNftTokens.map((nftItem) => ({
      chainId: nftItem.chainId,
      contractAddress,
      tokenId: nftItem.nftTokenId,
    })),
  );

  return useMemo(
    () => ({
      isLoadTokenError,
      nftCollectionItems: filteredNftTokens.map((token) => ({
        ...token,
        metadata: (nftItemMetadata ?? []).find((metadata) => metadata.tokenId === token.nftTokenId),
      })),
      isLoading: isLoadingTokenData || isLoadingMetadata,
    }),
    [filteredNftTokens, isLoadingMetadata, isLoadingTokenData, nftItemMetadata, isLoadTokenError],
  );
};
