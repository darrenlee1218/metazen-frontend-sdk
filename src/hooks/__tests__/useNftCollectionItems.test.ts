import NftAssetMetadata from '@gryfyn-types/data-transfer-objects/NftAssetMetadata';
import NftTokenData from '@gryfyn-types/data-transfer-objects/NftTokenData';
import { useNftCollectionItems } from '@hooks/useNftCollectionItems';
import { useGetAssetMetadataBatchQuery } from '@redux/api/assetMetadataCache';
import { useGetBalancesByAssetIdsQuery } from '@redux/api/bookKeeping';
import { renderHook } from '@testing-library/react-hooks';
import { mockImport } from '@utils/testing/mockImport';
import { constants } from 'ethers';

const mockNftTokens = [
  {
    chainId: '0',
    contractAddress: constants.AddressZero,
    nftTokenId: '1',
    balance: '1',
  },
  {
    chainId: '0',
    contractAddress: constants.AddressZero,
    nftTokenId: '2',
    balance: '1',
  },
  {
    chainId: '0',
    contractAddress: constants.AddressZero,
    nftTokenId: '3',
    balance: '1',
  },
  {
    chainId: '0',
    contractAddress: constants.AddressZero,
    nftTokenId: '4',
    balance: '0',
  },
] as unknown as NftTokenData[];

const mockNftAssetMetadata = [
  {
    tokenId: '1',
  },
  {
    tokenId: '2',
  },
] as unknown as NftAssetMetadata[];

jest.mock('@redux/api/bookKeeping');
const mockUseGetBalancesByAssetIdsQuery = mockImport(useGetBalancesByAssetIdsQuery);
jest.mock('@redux/api/assetMetadataCache');
const mockUseGetAssetMetadataBatchQuery = mockImport(useGetAssetMetadataBatchQuery);

it('should return an empty array if nftTokens is an empty array', () => {
  mockUseGetBalancesByAssetIdsQuery.mockReturnValue({ data: [] } as unknown as any);
  mockUseGetAssetMetadataBatchQuery.mockReturnValue({ data: mockNftAssetMetadata } as unknown as any);
  const { result } = renderHook(() =>
    useNftCollectionItems({ id: '', contractAddress: '0x0000000000000000000000000000000000000000' }),
  );
  expect(result.current.nftCollectionItems).toHaveLength(0);
});

it('should filter out items with no balance', () => {
  mockUseGetBalancesByAssetIdsQuery.mockReturnValue({ data: mockNftTokens } as unknown as any);
  mockUseGetAssetMetadataBatchQuery.mockReturnValue({ data: mockNftAssetMetadata } as unknown as any);
  const { result } = renderHook(() =>
    useNftCollectionItems({ id: '', contractAddress: '0x0000000000000000000000000000000000000000' }),
  );

  expect(result.current.nftCollectionItems.find((item) => item.nftTokenId === '4')).toBeUndefined();
});

it('should merge with nftMetadata', () => {
  mockUseGetBalancesByAssetIdsQuery.mockReturnValue({ data: mockNftTokens } as unknown as any);
  mockUseGetAssetMetadataBatchQuery.mockReturnValue({ data: mockNftAssetMetadata } as unknown as any);
  const { result } = renderHook(() =>
    useNftCollectionItems({ id: '', contractAddress: '0x0000000000000000000000000000000000000000' }),
  );
  expect(result.current).toMatchSnapshot();
});

test.each`
  loadingTokens | loadingMetadata | expected
  ${true}       | ${true}         | ${true}
  ${true}       | ${false}        | ${true}
  ${false}      | ${true}         | ${true}
  ${false}      | ${false}        | ${false}
`(
  'returns $expected when loadingTokens is $loadingTokens and loadingMetadata is $loadingMetadata',
  ({ loadingMetadata, loadingTokens, expected }) => {
    mockUseGetBalancesByAssetIdsQuery.mockReturnValue({ data: undefined, isLoading: loadingTokens } as unknown as any);
    mockUseGetAssetMetadataBatchQuery.mockReturnValue({
      data: undefined,
      isLoading: loadingMetadata,
    } as unknown as any);

    const { result } = renderHook(() =>
      useNftCollectionItems({ id: '', contractAddress: '0x0000000000000000000000000000000000000000' }),
    );
    expect(result.current.isLoading).toBe(expected);
  },
);
