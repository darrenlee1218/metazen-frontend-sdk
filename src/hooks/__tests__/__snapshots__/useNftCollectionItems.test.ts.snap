// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should merge with nftMetadata 1`] = `
Object {
  "isLoadTokenError": undefined,
  "isLoading": undefined,
  "nftCollectionItems": Array [
    Object {
      "balance": "1",
      "chainId": "0",
      "contractAddress": "0x0000000000000000000000000000000000000000",
      "metadata": Object {
        "tokenId": "1",
      },
      "nftTokenId": "1",
    },
    Object {
      "balance": "1",
      "chainId": "0",
      "contractAddress": "0x0000000000000000000000000000000000000000",
      "metadata": Object {
        "tokenId": "2",
      },
      "nftTokenId": "2",
    },
    Object {
      "balance": "1",
      "chainId": "0",
      "contractAddress": "0x0000000000000000000000000000000000000000",
      "metadata": undefined,
      "nftTokenId": "3",
    },
  ],
}
`;
